libpod/container_api.go

```diff
@@ -125,84 +166,80 @@ func (c *Container) Start(ctx context.Context) (err error) {
-// sent
-func (c *Container) StartAndAttach(ctx context.Context, streams *AttachStreams, keys string, resize <-chan remotecommand.TerminalSize) (attachResChan <-chan error, err error) {
+// RestartWithTimeout restarts a running container and takes a given timeout in uint
+func (c *Container) RestartWithTimeout(ctx context.Context, timeout uint) (err error) {
 	if !c.batched {
 		c.lock.Lock()
 		defer c.lock.Unlock()
 
 		if err := c.syncContainer(); err != nil {
-			return nil, err
+			return err
 		}
 	}
 
-	// Container must be created or stopped to be started
-	if !(c.state.State == ContainerStateConfigured ||
-		c.state.State == ContainerStateCreated ||
-		c.state.State == ContainerStateStopped ||
-		c.state.State == ContainerStateExited) {
-		return nil, errors.Wrapf(ErrCtrStateInvalid, "container %s must be in Created or Stopped state to be started", c.ID())
+	if err = c.checkDependenciesAndHandleError(ctx); err != nil {
+		return err
 	}
 
+	return c.restartWithTimeout(ctx, timeout)
+}
+
+// helper function that handles dependencies and prints a helpful message
+func (c *Container) checkDependenciesAndHandleError(ctx context.Context) error {
 	notRunning, err := c.checkDependenciesRunning()
 	if err != nil {
-		return nil, errors.Wrapf(err, "error checking dependencies for container %s", c.ID())
+		return errors.Wrapf(err, "error checking dependencies for container %s", c.ID())
 	}
 	if len(notRunning) > 0 {
 		depString := strings.Join(notRunning, ",")
-		return nil, errors.Wrapf(ErrCtrStateInvalid, "some dependencies of container %s are not started: %s", c.ID(), depString)
+		return errors.Wrapf(ErrCtrStateInvalid, "some dependencies of container %s are not started: %s", c.ID(), depString)
 	}
 
-	defer func() {
-		if err != nil {
-			if err2 := c.cleanup(ctx); err2 != nil {
-				logrus.Errorf("error cleaning up container %s: %v", c.ID(), err2)
-			}
-		}
-	}()
+	return nil
+}
 
-	if err := c.prepare(); err != nil {
-		return nil, err
+func (c *Container) startDependencies(ctx context.Context) error {
+	depCtrIDs := c.Dependencies()
+	if len(depCtrIDs) == 0 {
+		return nil
 	}
 
-	if c.state.State == ContainerStateStopped {
-		// Reinitialize the container if we need to
-		if err := c.reinit(ctx); err != nil {
-			return nil, err
-		}
-	} else if c.state.State == ContainerStateConfigured ||
-		c.state.State == ContainerStateExited {
-		// Or initialize it if necessary
-		if err := c.init(ctx); err != nil {
-			return nil, err
+	depCtrs := make([]*Container, 0) //, len(depCtrIDs))
+	for _, ctrID := range depCtrIDs {
+		ctr, err := c.runtime.state.LookupContainer(ctrID)
+		if err != nil {
+			return errors.Wrapf(err, "error starting dependency for container %s. container %s not found", c.ID(), ctrID)
 		}
+		depCtrs = append(depCtrs, ctr)
 	}
 
-	attachChan := make(chan error)
+	// Build a dependency graph of containers
+	graph, err := buildContainerGraph(depCtrs)
```

Comment:
I don't know how well this works if the dependencies have dependencies - it was originally written assuming every container it would ever touch would be passed in as part of the `[]*Container`

Reply:
ah shoot yeah I forgot to mention that. I haven't tested myself but I'm pretty sure things would get funky. Looking into better solution...